client/consensus/aura/src/lib.rs

```diff
@@ -264,11 +272,30 @@ impl<B, C, E, I, P, Error, SO> sc_consensus_slots::SimpleSlotWorker<B> for AuraW
 		StorageChanges<sp_api::TransactionFor<C, B>, B>,
 		Self::Claim,
 		Self::EpochData,
-	) -> sp_consensus::BlockImportParams<B, sp_api::TransactionFor<C, B>> + Send> {
-		Box::new(|header, header_hash, body, storage_changes, pair, _epoch| {
+	) -> Result<sp_consensus::BlockImportParams<B, sp_api::TransactionFor<C, B>>, sp_consensus::Error> + Send + 'static> {
+		let keystore = self.keystore.clone();
+		Box::new(move |header, header_hash, body, storage_changes, pair, _epoch| {
 			// sign the pre-sealed hash of the block and then
 			// add it to a digest item.
-			let signature = pair.sign(header_hash.as_ref());
+			let public = pair.public().to_raw_vec();
+			let public_type_pair = CryptoTypePublicPair(
+				<AuthorityId<Self::Claim> as RuntimeAppPublic>::CRYPTO_ID,
+				pair.public().to_raw_vec()
+			);
+			let signature = keystore.read()
+				.sign_with(
+					<AuthorityId<Self::Claim> as RuntimeAppPublic>::ID,
```

Comment:
Hmm fair enough... but how do i get access to `ID` and `CRYPTO_ID` when not using `RuntimeAppPublic`?

Reply:
`AppPublic`?

Reply:
Ok i see your point. I made the PR #5965 for the exact purpose of using `CRYPTO_ID` here. I missed the fact that `AppKey` holds `ID`. I will introduce `CRYPTO_ID` for `AppKey` as well so that it's usable in `AppPublic`.

Reply:
PR: https://github.com/paritytech/substrate/pull/6011